### PR TITLE
Taboola Adapter: support native 

### PR DESCRIPTION
**Pull request by ahmadlob**

### CODE REVIEW

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:43]: There is no need to check this. Request always has at least one impression.

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:45]: Note that you determine whole request type just by the first impression. It may be a potential source of bugs if there are impressions with different media types in one request. However this is not a likely case. Optional suggestion: split request to sub requests if there are imps with different types and use corresponding  endpoint urls.

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:16]: No need to hardcode it here. Modify endpoint URL in taboola.yaml file to 
`endpoint: "https://{{.MediaType}}.bidder.taboola.com/OpenRTB/TaboolaHB/auction"` and replace `MediaType` when you build request url. Similar to https://github.com/prebid/prebid-server/blob/91408f7d797dc7b425543a93e2eac8c4f5e228a4/adapters/adview/adview.go#L141

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:52]: This can also be a part of "buildEndpointUrl" function

File: static/bidder-info/taboola.yaml
Comment (VeronikaSolovei9) [inline static/bidder-info/taboola.yaml:13]: Please update adapter documentation accordingly: https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/taboola.md

File: adapters/taboola/taboola.go
Comment (gargcreation1992) [reply adapters/taboola/taboola.go:45]: I would also like to request if you could add a test where request has multiple impressions having at least one native and banner bid type.

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: check removed

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @VeronikaSolovei9  @gargcreation1992 
we mentioned in the documentation that currently we don't support **multi-formate requests.** 
However, we already added such a test only for checking the adapter functionality with multi-format requests (our server will respond with no-bid)
the unit test file name is: `multiImpressionsRequest.json`

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go:14]: @VeronikaSolovei9  Thanks for the great suggestion! just committed the changes using macros from now and on for building the endpoint url.

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: done

File: static/bidder-info/taboola.yaml
Comment (ahmadlob) [reply static/bidder-info/taboola.yaml:13]: will do once we implement the changes in the Prebid.js adapter

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: `multiFormatImpressionsRequest.json`

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [reply adapters/taboola/taboola.go:14]: Thank you for the change! Code looks so much cleaner now!

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:119]: Optional improvement: when you execute `buildRequest` function you already know what type is it. Instead of L116-L119 just pass  "display" or "native" from `MakeRequests`. Also please extract "display" and "native" to constants

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:142]: Niice!

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [inline adapters/taboola/taboola.go:244]: Fancy!

**Review by VeronikaSolovei9 — COMMENTED** [review]
Added minor comment. Changes look good to me. Ready to approve this PR if you decide not to modify it anymore

File: static/bidder-info/taboola.yaml
Comment (VeronikaSolovei9) [reply static/bidder-info/taboola.yaml:13]: Need to confirm it with the team.

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @VeronikaSolovei9  You are totally right! 
just committed the change :)

File: adapters/taboola/taboola.go
Comment (gargcreation1992) [inline adapters/taboola/taboola.go:109]: this function does not need to be defined as function receiver. 
```
func buildRequest(request *openrtb2.BidRequest, mediaType string) (*adapters.RequestData, error)
```

Comment (ahmadlob) [reply adapters/taboola/taboola.go]: I calls another method so it should receive the adapter and to be a method 
`a.buildEndpointUR`

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: I did the change, moved the different creation to be within the `createTaboolaRequests` function 
now looping **_only one time_** over the impressions

**Review by VeronikaSolovei9 — APPROVED** [review]
Thank you for the changes. Please note that a documentation update PR may be needed before merging it.

File: adapters/taboola/taboola.go
Comment (gargcreation1992) [inline adapters/taboola/taboola.go:113]: This code will probably will not be reachable since core framework already checks for Bid type based on `.yaml` file. However, if you would like to include it then this line is potential bug since err = nil always instead I would suggest either to remove this else or do - 
```
return nil, fmt.Errorf("unsupported media type for imp: %v", request.Imp[0])
```

Comment (VeronikaSolovei9) [reply adapters/taboola/taboola.go:113]: Agree, L112 is not returned or assigned to a variable.

Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @gargcreation1992  @VeronikaSolovei9 

thanks, did the changes. 

Please will appreciate if you merge this into the next version, I updated the documentation 

https://github.com/prebid/prebid.github.io/pull/4329

**Review by gargcreation1992 — DISMISSED** [review]
LGTM!

**Review by VeronikaSolovei9 — APPROVED** [review]
Thank you, LGTM

**Review by VeronikaSolovei9 — APPROVED** [review]

File: adapters/taboola/taboola.go
Comment (onkarvhanumante) [inline adapters/taboola/taboola.go:25]: scope of `NATIVE_ENDPOINT_PREFIX` and `DISPLAY_ENDPOINT_PREFIX` is limited to taboola package. so we can make them private

```
const (
   nativeEndpointPrefix = "native"
   displayEndpointPrefix = "display"
)
```

File: adapters/taboola/taboola.go
Comment (onkarvhanumante) [inline adapters/taboola/taboola.go:120]: `adapter`  need to suffice `adapters.Bidder` interface. Therefore methods like MakeBids,  MakeRequests are defined on `adapter`.    However there is no such need for buildRequest.  So `buildRequest` doesn't need to be defined on `(a *adapter)`

```
func  buildRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
  ...
}
```

File: adapters/taboola/taboola.go
Comment (onkarvhanumante) [inline adapters/taboola/taboola.go:151]: `adapter` need to suffice `adapters.Bidder` interface. Therefore methods like MakeBids, MakeRequests are defined on adapter.

However there is no such need for buildEndpointURL. So buildEndpointURL doesn't have to be defined on `(a *adapter)`

```
func  buildEndpointURL(publisherId string, mediaType string) (string, error) {
...
}
```

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go:120]: @onkarvhanumante  It's accessing some struct properties that's why it is defined as a method. 
It can be set as a function and pass the struct properties as arguments.

File: adapters/taboola/taboola.go
Comment (onkarvhanumante) [inline adapters/taboola/taboola.go:263]: nitpicks:  typo here.  `evventrackers` should be `eventrackers`

Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @onkarvhanumante  fixed, thanks!

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @onkarvhanumante  did the changes as you suggested. there are no longer an adapter methods.

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @onkarvhanumante  did the changes

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @onkarvhanumante  added them to an internal scope

**Review by onkarvhanumante — DISMISSED** [review]

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [reply adapters/taboola/taboola.go]: May I reopen this discussion please? This change is not needed, refer to `buildEndpointURL` in other adapters, we expect this function to get a receiver.

Comment (SyntaxNode) [reply adapters/taboola/taboola.go]: `buildEndpointURL` can exist as a function receiver in this case to access the endpoint template.

Comment (ahmadlob) [reply adapters/taboola/taboola.go]: I appreciate each feedback you provide, we also aim towards a more clean efficient code and to obey Prebid convention. 
But this is going back and forth for a while, If you could write your all agree on changes that need to be done so once and for all we will do the require changes will be really great. 

this PR https://github.com/prebid/prebid-server/pull/2522 was on hold  till this one is merge, because I didn't think this PR will take this significant amount of time, so I just merged it to this PR currently. 

@SyntaxNode 
@VeronikaSolovei9 
@onkarvhanumante

File: adapters/taboola/taboola.go
Comment (onkarvhanumante) [inline adapters/taboola/taboola.go:38]: `err` on line 26 gets shadowed by `err` on line 30. This function can be rephrased as follow

```
endpointTemplate, err := template.New("endpointTemplate").Parse(config.Endpoint)
if err != nil {
    return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
}

parsedUrl, err := url.Parse(server.ExternalUrl)
if err != nil {
     return nil, fmt.Errorf("unable to parse server external url: %v", err)
}

bidder := &adapter{
    endpoint: endpointTemplate,
    hostName: parsedUrl.Host,
}
return bidder, nil

```

File: static/bidder-info/taboola.yaml
Comment (onkarvhanumante) [inline static/bidder-info/taboola.yaml:19]: I believe this should be related to discussion done [here](https://github.com/prebid/prebid-server/pull/2522/files#r1086900875).

redirect urls are not redirecting.

Screenshot 2023-02-08 at 9 16 03 PM
Screenshot 2023-02-08 at 9 15 42 PM

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [reply adapters/taboola/taboola.go]: I'm sorry about this. 
Please revert it to original `func (a *adapter) buildEndpointURL(publisherId string, mediaType string)`
I think general suggestion here is to use receiver functions where you need to access adapter fields.

Comment (ahmadlob) [reply adapters/taboola/taboola.go:163]: @VeronikaSolovei9  done :)

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go:38]: @onkarvhanumante  
added error return in case parse endpoint url failed. 

regarding the second check, it failed to extract hostName I pass empty hostName and don't want to fail the whole function. 

changes committed

**Review by VeronikaSolovei9 — APPROVED** [review]
Thank you for the changes, looks good

**Comment by VeronikaSolovei9** [comment]
bidderinfo.go, should be:
```
var testEndpointTemplateParams = macros.EndpointTemplateParams{
	Host:        "anyHost",
	PublisherID: "anyPublisherID",
	AccountID:   "anyAccountID",
	ZoneID:      "anyZoneID",
	SourceId:    "anySourceID",
	AdUnit:      "anyAdUnit",
	MediaType:   "MediaType", <----- this value here
}
```
It is done for another open adapter PR, for some reason I decided it's there already. 
Sorry, could you please add this, so this branch works? Also please rebase/merge your changes on top of latest upstream branch.

File: adapters/taboola/taboola.go
Comment (VeronikaSolovei9) [reply adapters/taboola/taboola.go]: Just to confirm. With empty host final url looks like this: 
"https://display.bidder.taboola.com/OpenRTB/PS/auction//abc"
Is this ok to have `//`?

**Comment by ahmadlob** [comment]
@VeronikaSolovei9  committed the change in `bidderinfo` and merged the latest upstream version into the branch, thank you!

**Comment by VeronikaSolovei9** [comment]
Great thank you for the changes! Did you have a chance to review this comment: https://github.com/prebid/prebid-server/pull/2550#discussion_r1101874645 
I discussed this with the team, it's ok to have it if it works and you are ok with this too

File: adapters/taboola/taboola.go
Comment (ahmadlob) [reply adapters/taboola/taboola.go]: @VeronikaSolovei9  yes - it's fine.

**Review by VeronikaSolovei9 — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Pull request merged**
